README
Apache2::AuthCookieDBI is a module that subclasses Apache2::AuthCookie and is
designed to be directly used for authentication in a mod_perl server.

It is a ticket-issuing system that looks up username/passwords in a DBI
database using generic SQL and issues MD5-checksummed tickets valid for
a configurable time period.  Incoming requests with tickets are
checksummed and expire-time checked.

Version 2.03 and later: mod_perl 1.999_22 and later. Apache::*
replaced by Apache2::

Latest version at: http://search.cpan.org/dist/Apache2-AuthCookieDBI

Apache::AuthCookieDBI versions:
Version 2.0 and later: mod_perl 1.99_XX
Version 1.22 was the last version that works with mod_perl 1.x


Maintainer: matisse@cpan.org


Changes
Revision history for Perl extension Apache2::AuthCookieDBI.

2.19 - {{TBD}}
     - Added support for group authorizations on Apache 2.4.x. This addresses
       https://github.com/matisse/Apache-AuthCookieDBI/issues/2 and
       https://rt.cpan.org/Public/Bug/Display.html?id=106663.
     - Refactored group authorization code.
     - Eliminated duplicate calls of _dbi_config_vars() to improve efficiency.
     - Added docker directory that contains Dockerfiles for Apache 2.2 and 2.4
       Changes by Ed Sabol https://github.com/esabol

2.18 - Sat Aug 17 12:35:38 PDT 2019
     - Fix https://github.com/matisse/Apache-AuthCookieDBI/issues/3
       "DBI_CryptType crypt does not appear to work"
       Changes by Ed Sabol https://github.com/esabol

2.17 - Thu Dec  6 03:41:41 GMT 2012
     - Added support for Digest::SHA::sha256/384/512_hex digests for passwords.
       This is a response to https://rt.cpan.org/Ticket/Display.html?id=79333
       which requested sha256_base64 but because base64 digests are not
       properly padded. I chose to go with sha256/384/512_hex instead.

     - Quote all database column and field names in SQL queries.
       https://rt.cpan.org/Ticket/Display.html?id=79341
       'Table column names may confict with sql key words'

2.16 - Sun May 27 00:23:41 PDT 2012
     - Fix 'wrong method names in calls to apache request logging' for https://rt.cpan.org/Ticket/Display.html?id=77464

2.15 - Sat May 26 08:32:23 PDT 2012
     - Implemented a logger() method that uses Apache::Log (inspired by https://rt.cpan.org/Ticket/Display.html?id=75675)
     - Address 'Error Message Inconsistency' https://rt.cpan.org/Ticket/Display.html?id=75725
     - Address 'If user is not active, it produces 2 errors' https://rt.cpan.org/Ticket/Display.html?id=75724
     - Added small amount of test coverage for authen_cred()

2.14 - Sat Mar 26 09:41:01 PDT 2011
     - Changed documentation about WhateverDBI_SecretKey to
       address https://rt.cpan.org/Ticket/Display.html?id=62488
     - Fixes for https://rt.cpan.org/Public/Bug/Display.html?id=66336
       (_dbi_connect should be called as a class method.)

2.13 - Sat Dec 18 10:33:19 PST 2010
     - Build system uses Module::Build (generates a Makefile.PL for compatability.)
     - Distro version now matches module version (2.13)
     - Fixed statement handle still active <https://rt.cpan.org/Public/Bug/Display.html?id=57906>
       Also added support for WhatEverDBI_UserActiveField
       Thanks to Chad Columbus
     - Fixed $HEX_STRING_REGEX definition is wrong <https://rt.cpan.org/Public/Bug/Display.html?id=46117>
       Thanks to Joe Ingersoll
     - Fixed User authenticated if not in MySQL database <https://rt.cpan.org/Ticket/Display.html?id=62470>
       Thanks to Keith Lawson
     - Fixed Login.pl example - syntax error <https://rt.cpan.org/Public/Bug/Display.html?id=62487>
       (also moved from eg/public-pl/login.pl to eg/bin/login.pl)
       Thanks to William McKee http://search.cpan.org/~wmckee/
     - Updated documentation to cover protecting the DocumentRoot <https://rt.cpan.org/Ticket/Display.html?id=51593>
     - Now depends on DBI 1.40 instead of Apache::DBI - <https://rt.cpan.org/Ticket/Display.html?id=43083>

2.05 - Thu May 14 18:37:07 PDT 2009
       - Fixed bug in _get_cipher_type() where it was not memoizing the
         Crypt::CBC objects.
       - Fixed https://rt.cpan.org/Ticket/Display.html?id=45207
         Hash keys for configuration values were wrong in several places.
       - Improved test coverage a little.

2.04 - Fri Nov 28 15:41:33 PST 2008
       Incorporated bug fix for authen_ses_key() provided by
       Carl Gustafsson. authen_ses_key() was not properly handling
       any extra_session_info - the fix is to get $hashed_string with
          my $hashed_string = pop @rest;

       Also releasing the work done between April 26, 2005 and February 4, 2007,
       plus bug fix for authen_ses_key
       - Added basic framework for unit tests.
       - Factored out some of the DBI code into new methods:
         - _dbi_connect()
         - _get_crypted_password()

       - Changes to satisfy Perl::Critic, e.g.
         - Removed function prototypes (they are ignored for methods.)
         - Cleaned up regular expressions: use /x, etc.

2.03 - Mon Apr 25 10:01:04 PDT 2005
       Incorporated Lance P Cleveland's changes porting module to mod_perl 1.999_22
       (That is, Version 2.0.0-RC5 - April 14, 2005)
       Mainly involves changing almost all references to Apache:: to Apache2::

       Also some minor documentation changes.

2.02 - Tue Apr 12 17:58:40 PDT 2005
       Minor documentation fix. Noted that SecretKeyFile has been
       replaced by SecretKey.

2.0.1- Fixed bug in group(). It was expecting the list of groups to be
       already split up.

2.0  - mod_perl 2 version

0.01 - Mon Apr  3 10:50:32 2000
       - original version; created by h2xs 1.19



t/mock_libs/DBI.pm
# $Header: /Users/matisse/Desktop/CVS2GIT/matisse.net.cvs/Apache-AuthCookieDBI/t/mock_libs/DBI.pm,v 1.3 2009/04/26 17:33:26 matisse Exp $
# $Revision: 1.3 $
# $Author: matisse $
# $Source: /Users/matisse/Desktop/CVS2GIT/matisse.net.cvs/Apache-AuthCookieDBI/t/mock_libs/DBI.pm,v $
# $Date: 2009/04/26 17:33:26 $
###############################################################################

#  Mock class - for testing only

package DBI;
use strict;
use warnings;

#warn 'Loading mock library ' . __FILE__;
my $MOCK_DBH_CLASS = 'DBI::Mock::dbh';

our $CONNECT_CACHED_FORCE_FAIL;

sub connect_cached {
    my ( $class, @args ) = @_;

    if ($CONNECT_CACHED_FORCE_FAIL) {
        return;
    }

    my $fake_dbh = {};
    bless $fake_dbh, $MOCK_DBH_CLASS;
    $fake_dbh->{'connect_cached_args'} = \@args;

    return $fake_dbh;
}

1;


t/mock_libs/Crypt/CBC.pm
package Crypt::CBC;
use strict;
use warnings;

sub new {
    my ( $class, $secret_key, $cipher_type ) = @_;
    my $self = {
        secret_key  => $secret_key,
        cipher_type => $cipher_type,
    };
    bless $self, $class;
    return $self;
}

sub encrypt_hex {
    my ( $self, $plain_text ) = @_;
    my $mock_crypt_text = join q{:}, $self->{cipher_type}, $self->{secret_key},$plain_text;
    return $mock_crypt_text;
}

1;


eg/public-pl/login.pl
#!/usr/bin/perl -Tw
#
# $Id: login.pl,v 1.4 2010/11/27 04:11:19 matisse Exp $
#
# Display a login form with hidden fields corresponding to the page they
# wanted to see.

use strict;
use 5.004;
use Text::TagTemplate;
use Apache;

my $t = new Text::TagTemplate;
my $r = Apache->request();

my $destination;
my $authcookiereason;
if ( $r->prev() ) { # we are called as a subrequest.
	$destination = $r->prev()->args()
	             ? $r->prev()->uri() . '?' .  $r->prev->args()
	             : $r->prev()->uri();
	$authcookiereason = $r->prev()->subprocess_env( 'AuthCookieReason' );
} else {
        my %args = $r->args;
	$destination = $args{ 'destination' };
	$authcookiereason = $args{ 'AuthCookieReason' };
        $t->add_tag( CREDENTIAL_0 => $r->prev->args('credential_0') );
}
$t->add_tag( DESTINATION => $destination );

unless ( $authcookiereason eq 'bad_cookie' ) {
	$t->template_file( "../html/login.html" );
} else {
	$t->template_file( "../html/login-failed.html" );
}

$r->send_http_header;
print $t->parse_file unless $r->header_only;


lib/Apache2_4/AuthCookieDBI.pm
#===============================================================================
#
# Apache2_4::AuthCookieDBI
#
# A module implementing Apache 2.4.x compatibility for Apache2::AuthCookieDBI
# group-based authorizations.
#
# See end of this file for Copyright notices.
#
# Maintainer: Matisse Enzer <matisse@cpan.org> (as of version 2.0)
#
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#===============================================================================

package Apache2_4::AuthCookieDBI;

# ABSTRACT: Perl group authorization for Apache 2.4.x

use strict;
use warnings;
use base 'Apache2::AuthCookieDBI';
use Apache2::Log;
use Apache2::Const -compile => qw(AUTHZ_GRANTED AUTHZ_DENIED AUTHZ_DENIED_NO_USER AUTHZ_GENERAL_ERROR);

#===============================================================================
# FILE (LEXICAL)  G L O B A L S
#===============================================================================

our $VERSION = '2.19';

#===============================================================================
# M E T H O D ( S )
#===============================================================================

sub group {
    my ($class, $r, $groups) = @_;

    my $debug = $r->dir_config('AuthCookieDebug') || 0;
    my $user = $r->user;

    $r->server->log_error("authz:start user=@{[ defined($user) ? $user : '(undef)' ]} type=$class groups=@{[ defined($groups) ? $groups : '(undef)' ]} uri=@{[ $r->uri ]}") if ($debug >= 5);

    if ( _is_empty($user) ) {
        # User is not yet authenticated.
        return Apache2::Const::AUTHZ_DENIED_NO_USER;
    }

    if ( _is_empty($groups) ) {
        my $message
            = "${class}\tno group(s) specified in the \'Require group ...\' configuration for URI @{[ $r->uri ]}";
        $class->logger( $r, Apache2::Const::LOG_INFO, $message, $user,
            $class->LOG_TYPE_AUTHZ, $r->uri );
        return Apache2::Const::AUTHZ_DENIED;
    }

    my @groups = split( $class->WHITESPACE_REGEX, $groups );

    # Instead of querying the database every time, check subprocess_env
    # for AUTH_COOKIE_DBI_GROUP (set below) and utilize that to short circuit
    # the authorization.
    if ( my $cached_group = $r->subprocess_env('AUTH_COOKIE_DBI_GROUP') ) {
        $r->server->log_error("${class}\tfound cached group $cached_group") if ($debug >= 3);
        foreach my $group (@groups) {
            if ( $group eq $cached_group ) {
                return Apache2::Const::AUTHZ_GRANTED;
            }
        }
    }

    $r->server->log_error("authz user=$user type=$class group=$groups") if ($debug >= 2);

    # Get the configuration information.
    my %c = $class->_dbi_config_vars($r);

    # See if we have a row in the groups table for this user/group.
    my $dbh = $class->_dbi_connect($r, \%c)
      || return Apache2::Const::AUTHZ_GENERAL_ERROR;
    my $sth = $class->_prepare_group_query($dbh, \%c)
      || return Apache2::Const::AUTHZ_GENERAL_ERROR;

    return $class->_check_group_membership($r, $sth, \@groups, $debug)
      ? Apache2::Const::AUTHZ_GRANTED
      : Apache2::Const::AUTHZ_DENIED;
}

1;

__END__

#===============================================================================
# P E R L D O C
#===============================================================================

=head1 NAME

Apache2_4::AuthCookieDBI - A subclass of L<Apache2::AuthCookieDBI>
that implements a "group" authorization provider for Apache 2.4.x.

=head1 SYNOPSIS

 # In httpd.conf or .htaccess:
 #   Configure as you would with Apache2::AuthCookieDBI, but leave out
 #   the PerlAuthzHandler directive and add the following when using
 #   "require group":

 PerlModule Apache2_4::AuthCookieDBI
 PerlAddAuthzProvider group Apache2_4::AuthCookieDBI->group

 <Location /www/domain.com/authcookiedbi/admin>
   require group admin
 </Location>

=head1 DESCRIPTION

B<Apache2_4::AuthCookieDBI> provides an Apache 2.4.x-compatible
authorization provider for handling "group" authorization
requirements.

This module is a subclass of L<Apache2::AuthCookieDBI>. All the
methods for Apache2::AuthCookieDBI still work for this module as well.
The only method that this class overrides is C<group>.

This module is for F<mod_perl> version 2 and Apache version 2.4.x. If
you are running Apache 2.0.0-2.2.x, refer to L<Apache2::AuthCookieDBI>.

Make sure your F<mod_perl> is at least 2.0.9, with StackedHandlers,
MethodHandlers, Authen, and Authz compiled in.

=head1 HISTORY

The implementation herein is based on L<Apache2::AuthCookieDBI>'s C<group>
method with heavy inspiration from the sample C<authz_handler> in
L<Apache2_4::AuthCookie> by Michael Schout. Huge thanks to Michael Schout
for his documentation on the changes to authorization under Apache 2.4.x.

=head1 COPYRIGHT

 Copyright (C) 2002 SF Interactive
 Copyright (C) 2003-2004 Jacob Davies
 Copyright (C) 2004-2019 Matisse Enzer

=head1 LICENSE

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=head1 CREDITS

  Original Author: Jacob Davies
  Incomplete list of additional contributors (alphabetical by first name):
    Carl Gustafsson
    Chad Columbus
    Edward J Sabol
    Jay Strauss
    Joe Ingersoll
    Keith Lawson
    Lance P Cleveland
    Matisse Enzer
    Nick Phillips
    William McKee
      
=head1 MAINTAINER

Matisse Enzer

        <matisse@cpan.org>
        
=head1 SEE ALSO

 Latest version: http://search.cpan.org/dist/Apache2-AuthCookieDBI

 Apache2::AuthCookie - http://search.cpan.org/dist/Apache2-AuthCookie
 Apache2::Session    - http://search.cpan.org/dist/Apache2-Session
 Apache::AuthDBI     - http://search.cpan.org/dist/Apache-DBI

=cut


Build.PL
use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
    module_name => 'Apache2::AuthCookieDBI',
    license     => 'lgpl',
    requires    => {
        'mod_perl2'           => 1.999022,
        'Apache2::AuthCookie' => 3.08,
        'DBI'                 => 1.40,
        'Apache2::Const'      => 0,
        'Apache2::RequestRec' => 0,
        'Apache2::ServerUtil' => 0,
        'Carp'                => 0,
        'Crypt::CBC'          => 2.13,
        'Date::Calc'          => 0,
        'Digest::MD5'         => 0,
        'FindBin'             => 0,
    },
    build_requires     => { 'Test::More' => 0.4, },
    create_makefile_pl => 'traditional',
    add_to_cleanup     => [
        qw( *.tar.gz
            MANIFEST.bak
            ),
    ],
    meta_merge => {
        'author'    => [ 'Matisse Enzer <matisse@cpan.org>', ],
        'meta-spec' => {
            'version' => 2,
            'url' =>
                'http://search.cpan.org/perldoc?CPAN::Meta::Spec',
        },
        'release_status' => 'stable',
        'resources'      => {
            'license'    => ['http://opensource.org/licenses/lgpl-license.php'],
            'homepage'   => 'http://search.cpan.org/dist/Apache2-AuthCookieDBI',
            'bugtracker' => {
                'web' =>
                    'https://rt.cpan.org/Public/Dist/Display.html?Name=Apache2-AuthCookieDBI',
            },
            'repository' => {
                'url'  => 'git://github.com/matisse/Apache-AuthCookieDBI.git',
                'web'  => 'https://github.com/matisse/Apache-AuthCookieDBI',
                'type' => 'git',
            },
        },
    },

);

$build->create_build_script();

exit;
__END__

# 'ORIGINAL_AUTHOR'       => 'Jacob Davies (jacob@well.com)',
  
